src/core/communication.cpp

```diff
@@ -335,14 +335,15 @@ void mpi_bcast_ia_params_slave(int i, int j) {
     boost::mpi::broadcast(comm_cart, *get_ia_param(i, j), 0);
   } else {                   /* bonded interaction parameters */
     make_bond_type_exist(i); /* realloc bonded_ia_params on slave nodes! */
+    if (is_tabulated_bond(bonded_ia_params[i].type)) {
+      delete bonded_ia_params[i].p.tab.pot;
+    }
     MPI_Bcast(&(bonded_ia_params[i]), sizeof(Bonded_ia_parameters), MPI_BYTE, 0,
               comm_cart);
     /* For tabulated potentials we have to send the tables extra */
     if (is_tabulated_bond(bonded_ia_params[i].type)) {
-      auto *tab_pot = new TabulatedPotential();
-      boost::mpi::broadcast(comm_cart, *tab_pot, 0);
-
-      bonded_ia_params[i].p.tab.pot = tab_pot;
+      bonded_ia_params[i].p.tab.pot = new TabulatedPotential();
```

Comment:
haven't looked into the code, but can't we use smart pointers?

Reply:
I don't think so: we overwrite the memory block of the bond object on worker nodes with a bytestring of the same object from the head node, so the counter from a `shared_ptr` would be overwritten. As for `unique_ptr`, in the line you highlighted, we do an assignment, so the smart pointer goes out of scope and gets garbage-collected... but it contains a random memory address (because of the bytestring deserialization a few lines above), which is a recipe for segmentation faults.

Reply:
i'm not sure i'm following... but it sounds like this is a place for larger scale refactoring which is out of scope of this pull request?

Reply:
the correct way would be to write boost serialization methods, which can handle pointer members out of the box, e.g.
```c++
struct Tabulated_bond_parameters {
  int type;
  TabulatedPotential *pot;

private:
  friend boost::serialization::access;
  template <typename Archive>
  void serialize(Archive &ar, long int /* version */) {
    ar &type;
    ar &pot;
  }
};
```
but there is no guarantee that we are overwriting a bond type with the same bond type, so we would have to add destructors too... Not sure if this is worth the effort, if our long-term goal is to replace this interaction communication infrastructure with a python setter.

Reply:
Also, a much more promising refactoring goal is breaking the dependency on `MpiCallbacks.hpp` from consumers of MPI callbacks ([fix-3906...fix-3907-p1](https://github.com/jngrad/espresso/compare/fix-3906...fix-3907-p1)), which as a side effect is removing global variables and removing prototypes of MPI worker functions from header files.

Reply:
> Also, a much more promising refactoring goal is breaking the dependency on `MpiCallbacks.hpp` from consumers of MPI callbacks ([fix-3906...fix-3907-p1](https://github.com/jngrad/espresso/compare/fix-3906...fix-3907-p1)), which as a side effect is removing global variables and removing prototypes of MPI worker functions from header files.

Isn't that totally orthogonal to the memory leak issue?

Reply:
> the correct way would be to write boost serialization methods, which can handle pointer members out of the box, e.g.
> 
> ```c++
> struct Tabulated_bond_parameters {
>   int type;
>   TabulatedPotential *pot;
> 
> private:
>   friend boost::serialization::access;
>   template <typename Archive>
>   void serialize(Archive &ar, long int /* version */) {
>     ar &type;
>     ar &pot;
>   }
> };
> ```
> 
> but there is no guarantee that we are overwriting a bond type with the same bond type, so we would have to add destructors too... Not sure if this is worth the effort, if our long-term goal is to replace this interaction communication infrastructure with a python setter.

sorry, if that has been discussed and i don't remember it anymore... Do we have some design plans for the interaction stuff already?